lib/capability/capability.go

```diff
@@ -0,0 +1,94 @@
+package capability
+
+import (
+	"fmt"
+	"strings"
+
+	configv1 "github.com/openshift/api/config/v1"
+)
+
+const CapabilityAnnotation = "capability.openshift.io/name"
+
+type ClusterCapabilities struct {
+	KnownCapabilities map[configv1.ClusterVersionCapability]struct{}
+
+	// value indicates whether capability was explicitly included
+	EnabledCapabilities map[configv1.ClusterVersionCapability]bool
+}
+
+func SetCapabilities(config *configv1.ClusterVersion) ClusterCapabilities {
+	var capabilities ClusterCapabilities
+	capabilities.KnownCapabilities = setKnownCapabilities(config)
+	capabilities.EnabledCapabilities = setEnabledCapabilities(config)
+	return capabilities
+}
+
+func GetCapabilitiesStatus(capabilities ClusterCapabilities) configv1.ClusterVersionCapabilitiesStatus {
+	var status configv1.ClusterVersionCapabilitiesStatus
+	for k, _ := range capabilities.EnabledCapabilities {
+		status.EnabledCapabilities = append(status.EnabledCapabilities, k)
+	}
+	for k, _ := range capabilities.KnownCapabilities {
+		status.KnownCapabilities = append(status.KnownCapabilities, k)
+	}
+	return status
+}
+
+func CheckResourceEnablement(annotations map[string]string, capabilities ClusterCapabilities) error {
+	val, ok := annotations[CapabilityAnnotation]
+	if ok {
+		caps := strings.Split(val, "+")
+		numCaps := len(caps)
+		unknownCaps := make([]string, numCaps)
+		disabledCaps := make([]string, numCaps)
+
+		for _, c := range caps {
+			trimmedCap := strings.TrimSpace(c)
+
+			if _, ok = capabilities.KnownCapabilities[configv1.ClusterVersionCapability(trimmedCap)]; !ok {
+				unknownCaps = append(unknownCaps, trimmedCap)
+			} else if _, ok = capabilities.EnabledCapabilities[configv1.ClusterVersionCapability(trimmedCap)]; !ok {
+				disabledCaps = append(disabledCaps, trimmedCap)
+			}
+		}
+		if len(unknownCaps) > 0 {
+			return fmt.Errorf("unrecognized capability names: %s", strings.Join(unknownCaps, ", "))
+		}
+		if len(disabledCaps) > 0 {
+			return fmt.Errorf("disabled capabilities: %s", strings.Join(disabledCaps, ", "))
+		}
+	}
+	return nil
+}
+
+func setKnownCapabilities(config *configv1.ClusterVersion) map[configv1.ClusterVersionCapability]struct{} {
+	known := make(map[configv1.ClusterVersionCapability]struct{})
+
+	for _, v := range configv1.ClusterVersionCapabilitySets {
+		for _, capability := range v {
+			if _, ok := known[capability]; ok {
+				continue
+			}
+			known[capability] = struct{}{}
+			config.Status.Capabilities.KnownCapabilities = append(config.Status.Capabilities.KnownCapabilities, capability)
+		}
+	}
+	return known
+}
+
+func setEnabledCapabilities(config *configv1.ClusterVersion) map[configv1.ClusterVersionCapability]bool {
+	enabled := make(map[configv1.ClusterVersionCapability]bool)
+
+	for _, v := range configv1.ClusterVersionCapabilitySets[config.Spec.Capabilities.BaselineCapabilitySet] {
```

Comment:
[The API][1] docs `vCurrent` as the default cap set, but does not actually declare that default anywhere in an API property, so we'll need to define that defaulting locally here (in a way that we can easily factor it out into library-go later to support `oc adm release extract ...`).

[1]: https://github.com/openshift/api/blob/354aa98a475c1fcd60b41aaee735da95d7318b42/config/v1/types_cluster_version.go#L280-L281